nexus/src/app/background/abandoned_vmm_reaper.rs

```diff
@@ -0,0 +1,450 @@
+//!   fields).
+//! - It has not been deleted yet.
+
+use super::common::BackgroundTask;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_model::Vmm;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::pagination::Paginator;
+use nexus_db_queries::db::DataStore;
+use std::num::NonZeroU32;
+use std::sync::Arc;
+
+/// Background task that searches for abandoned VMM recordss and deletes them.
+pub struct AbandonedVmmReaper {
+    datastore: Arc<DataStore>,
+}
+
+#[derive(Debug, Default)]
+struct ActivationResults {
+    found: usize,
+    sled_reservations_deleted: usize,
+    vmms_deleted: usize,
+    vmms_already_deleted: usize,
+    error_count: usize,
+}
+
+const MAX_BATCH: NonZeroU32 = unsafe {
+    // Safety: last time I checked, 100 was greater than zero.
+    NonZeroU32::new_unchecked(100)
+};
+
+impl AbandonedVmmReaper {
+    pub fn new(datastore: Arc<DataStore>) -> Self {
+        Self { datastore }
+    }
+
+    /// List abandoned VMMs and clean up all of their database records.
+    async fn reap_all(
+        &mut self,
+        results: &mut ActivationResults,
+        opctx: &OpContext,
+    ) -> Result<(), anyhow::Error> {
+        slog::info!(opctx.log, "Abandoned VMM reaper running");
+
+        let mut paginator = Paginator::new(MAX_BATCH);
+        let mut last_err = Ok(());
+        while let Some(p) = paginator.next() {
+            let vmms = self
+                .datastore
+                .vmm_list_abandoned(opctx, &p.current_pagparams())
+                .await
+                .context("failed to list abandoned VMMs")?;
+            paginator = p.found_batch(&vmms, &|vmm| vmm.id);
+            self.reap_batch(results, &mut last_err, opctx, &vmms).await;
+        }
+
+        last_err
+    }
+
+    /// Clean up a batch of abandoned VMMs.
+    ///
+    /// This is separated out from `reap_all` to facilitate testing situations
+    /// where we race with another Nexus instance to delete an abandoned VMM. In
+    /// order to deterministically simulate such cases, we have to perform the
+    /// query to list abandoned VMMs, ensure that the VMM record is deleted, and
+    /// *then* perform the cleanup with the stale list of abandoned VMMs, rather
+    /// than doing it all in one go. Thus, this is factored out.
+    async fn reap_batch(
+        &mut self,
+        results: &mut ActivationResults,
+        last_err: &mut Result<(), anyhow::Error>,
+        opctx: &OpContext,
+        vmms: &[Vmm],
+    ) {
+        results.found += vmms.len();
+        slog::debug!(opctx.log, "Found abandoned VMMs"; "count" => vmms.len());
+
+        for vmm in vmms {
+            let vmm_id = vmm.id;
+            slog::trace!(opctx.log, "Deleting abandoned VMM"; "vmm" => %vmm_id);
+            // Attempt to remove the abandoned VMM's sled resource reservation.
+            match self.datastore.sled_reservation_delete(opctx, vmm_id).await {
```

Comment:
The one spot of this PR that feels weird to me is the overlap with functionality in https://github.com/oxidecomputer/omicron/blob/c2f35151c7ac345305f396bc01f088083ccae7e0/nexus/src/app/instance.rs#L1970 

When a VMM is destroyed, that function seems to:

- Update counters in the project / fleet about "virtual provisioning" (see: `virtual_provisioning_collection_delete_instance`)
- Unassign producers from oximeter (see: `unassign_producer`)
- Update instance and VMM states to indicate that the VMM has been destroyed
- Delete the sled reservation (physical usage)
- Mark the VMM deleted

(We have previously identified that it is a problem these steps aren't atomic! But that's an issue which exists independently of this PR)

In contrast, this background task seems to:

- Delete the sled reservation (physical usage)
- Mark the VMM deleted

So, I have a smattering of follow-up questions here:

- Should this background task take over this responsibility for destroying VMMs from `notify_instance_updated` ? e.g., should that function start kicking this background task?
- That function also seems to destroy VMMs in the `Failed` state. Is that scope we would want this task to handle too?
- Should the "physical usage" tracking (sled reservation) and "virtual usage" tracking (`virtual_provisioning_collection...`) updates be more tightly coupled? I think they're both already atomic.

Reply:
I *think* our discussion in the hypervisor sync and on Matrix today has cleared this up --- for readers other than @smklein, though, the goal is that eventually:

- The behavior currently run in `notify_instance_updated` will eventually move to a saga (see #5749), and the Nexus upcall will just update the instance's state and kick off that saga
- When a VMM that's _actively in use by an instance_ is destroyed/failed, a bunch of cleanup needs to happen, including virtual provisioning, Oximeter producer, and networking; the update saga will do this
- When a VMM is abandoned after a migration (either because it was a failed migration target, or because it was the old VMM leftover after a successful migration), the cleanup is much simpler, as the _instance_ is still around and still using the virtual provisioning resources. Just the physical sled resources consumed by the individual Propolis process need to be released, the other resources are still in use by the instance, just not via *this* VMM.

This background task is only responsible for abandoned VMMs, so it only needs to do the simple cleanup process for them; it doesn't deallocate resources owned by instances when an in-use VMM fails or is stopped.